src/table.js

```diff
@@ -538,15 +543,99 @@ export function getTypeValidator(colType) {
+  const numberDefault = defaultValue === null ? NaN : defaultValue;
+  switch (type) {
+    case "string":
+      return value === "string"
+        ? value.trim()
+        : value || value === 0
+        ? value.toString()
+        : defaultValue;
+    case "boolean":
+      return value === true || value === "true"
+        ? true
+        : value === false || value === "false"
+        ? false
+        : defaultValue;
+    case "integer":
+      return !value || isNaN(parseInt(value)) ? numberDefault : parseInt(value);
+    case "bigint":
+      return typeof value === "bigint"
+        ? value
+        : !value || isNaN(value)
+        ? numberDefault
+        // eslint-disable-next-line no-undef
+        : BigInt(value);
+    case "number": {
+      return !value || isNaN(value) ? numberDefault : Number(value);
+    }
+    case "date": {
+      if (value instanceof Date) return value;
+      if (typeof value === "string") {
+        let match;
+        if (
+          (match = value.match(
+            /^([-+]\d{2})?\d{4}(-\d{2}(-\d{2})?)?(T\d{2}:\d{2}(:\d{2}(\.\d{3})?)?(Z|[-+]\d{2}:\d{2})?)?$/
+          ))
+        ) {
+          if (fixTz && !!match[4] && !match[7])
+            value = value.replace(/-/g, "/").replace(/T/, " ");
+        }
+      }
+      // Invalid Date objects are still instances of Date, but they return true
+      // from isNaN(). If we are "soft-coercing," we want to return the original
+      // value for invalid dates. Otherwise, if a date is invalid, return an
+      // Invalid Date object.
+      const date = new Date(value);
+      const dateDefault = options.soft ? value : date;
+      return isNaN(date) ? dateDefault : date;
+    }
+    case "array":
+      if (Array.isArray(value)) return value;
+      return value && Array.isArray(Array.from(value))
+        ? Array.from(value)
+        : defaultValue;
+    case "buffer":
+    case "object":
+    case "other":
+    default:
+      return value || value === 0 ? value : defaultValue;
+  }
+}
+
 // This function applies table cell operations to an in-memory table (array of
 // objects); it should be equivalent to the corresponding SQL query. TODO Use
 // DuckDBClient for data arrays, too, and then we wouldn’t need our own __table
 // function to do table operations on in-memory data?
 export function __table(source, operations) {
   const input = source;
   let {schema, columns} = source;
+  let newlyInferred = false;
+  if (!schema || !isQueryResultSetSchema(schema)) {
+    schema = inferSchema(source);
+    newlyInferred = true;
+  }
   let primitive = arrayIsPrimitive(source);
   if (primitive) source = Array.from(source, (value) => ({value}));
+  // Combine column types from schema with user-selected types in operations
+  const types = new Map(schema.map(({name, type}) => [name, type]));
+  if (operations.type) {
+    for (const {name, type} of operations.type) {
+      types.set(name, type);
+      source = source.map(d => coerceRow(d, types));
```

Comment:
Not sure if this is a performance concern, but here's we're iterating through the entire dataset for each user-specified type. Could we iterate through the all the types and perform `types.set(name, type);`, then afterward just iterate through `source = source.map(d => coerceRow(d, types));` one time?